lib/src/rsa.nr

```diff
@@ -48,6 +114,113 @@ fn compare_signature_sha256<let N: u32>(padded_sha256_hash: [u8; N], msg_hash: [
+        key_size: u32
+    ) -> bool {
+        // Exponentiate the signature assuming e = 65537
+        let mut exponentiated = instance.mul(sig, sig);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, sig);
+        // Convert the exponentiated signature to a byte array and reverse it to 
+        // get it in big endian order, which is much easier to work with for 
+        // the rest of the verification process
+        let em:[u8; NumBytes] = reverse_array(exponentiated.to_le_bytes());
+
+        // The modulus size in bits minus 1
+        let em_bits = key_size - 1;
+        // The actual length of the encoded message without any of the leftmost 0s
+        let em_len = (em_bits + 7) / 8;
+        // The length of the modulus in bytes
+        let key_len = (key_size + 7) / 8;
+        let h_len = 32;
+        let s_len = 32;
+
+        // Check if emLen < hLen + sLen + 2
+        assert(em_len >= h_len + s_len + 2);
+
+        // Check if eM ends with 0xBC
+        assert(em[em.len() - 1] == 0xBC);
+
+        let db_mask_len = em_len - h_len - 1;
+        // This offset is necessary for key sizes not divisible by 8
+        // Will equal 0 for key sizes divisible by 8
+        let offset = em.len() - key_len;
+        // 256 - 32 - 1 = 223
+        // As hash is 32 bytes and we also remove the 0xBC at the end, we have 223 bytes left for DB
+        let masked_db: [u8; 223] = get_array_slice(em, offset, db_mask_len + offset);
+        let h = get_array_slice(em, db_mask_len + offset, em.len() - 1);
+
+        // Mask the leftmost bits
+        let bits_to_mask = 8 * em_len - em_bits;
+        let mask = 0xFF >> bits_to_mask as u8;
+        let first_byte = masked_db[offset];
+        // Make sure the 8 * em_len - em_bits leftmost bits are 0
+        // c.f. https://github.com/RustCrypto/RSA/blob/aeedb5adf5297892fcb9e11f7c0f6c0157005c58/src/algorithms/pss.rs#L205
+        assert((first_byte & !mask) == 0);
```

Comment:
This can be done more cheaply by just performing a division rather than bitwise operations.

Reply:
Indeed, I switched to using powers of 2 now